samza-kafka/src/main/scala/org/apache/samza/system/kafka/KafkaSystemProducer.scala

```diff
@@ -47,22 +39,13 @@ class KafkaSystemProducer(systemName: String,
 
   class SourceData {
     /**
-     * lock to make send() and store its future atomic
-     */
-    val sendLock: Object = new Object
-    /**
-     * The most recent send's Future handle
-     */
-    @volatile
-    var latestFuture: Future[RecordMetadata] = null
-    /**
      * exceptionInCallback: to store the exception in case of any "ultimate" send failure (ie. failure
      * after exhausting max_retries in Kafka producer) in the I/O thread, we do not continue to queue up more send
      * requests from the samza thread. It helps the samza thread identify if the failure happened in I/O thread or not.
      *
      * In cases of multiple exceptions in the callbacks, we keep the first one before throwing.
      */
-    var exceptionInCallback: AtomicReference[SamzaException] = new AtomicReference[SamzaException]()
+    var exceptionInCallback: AtomicReference[SystemProducerException] = new AtomicReference[SystemProducerException]()
```

Comment:
QQ: Does this need to be a var? Maybe make a val to avoid accidentally nulling the reference instead of calling getAndSet(null).